File: #100daysofcodewithGFG/Ruling Pair.py
class Solution:
    def RulingPair(self, arr, n): 
    	# Your code goes here
    	def func(item):
    	    su = 0
    	    while item>0:
    	        rem = item%10
    	        su+=rem
    	        item = item//10
    	    return su
    	       
    	from collections import defaultdict
    	dic = defaultdict(list)
    	for item in arr:
    	    ans = func(item)
    	    if len(dic[ans]) < 2:
    	        dic[ans].append(item)
    	    else:
    	        if item>dic[ans][0] and item>dic[ans][1]:
    	            second = max(dic[ans][0],dic[ans][1])
    	            dic[ans][0] = item
    	            dic[ans][1] = second
    	        elif item>dic[ans][0]:
    	            dic[ans][0] = item
    	        elif item>dic[ans][1]:
    	            dic[ans][1] = item
    	ma = -1
    	for k,v in dic.items():
    	    if len(v) == 2:
    	        one = v[0]
    	        two = v[1]
    	        ma = max(ma,one+two)
    	return ma

File: #100daysofcodewithGFG/Ruling Pair.cpp
class Solution{
    
    public:
    int func(int x){
        int su = 0;
        while(x>0){
            int rem = x%10;
            su+=rem;
            x = x/10;
        }
        return su;
    }
    int RulingPair(vector<int> arr, int n) 
    { 
    	// Your code goes here
    	unordered_map<int,int> map;
    	int ans = -1;
    	for(int i=0;i<n;i++){
    	    int sum = func(arr[i]);
    	    if (map.find(sum) == map.end()){
    	        map[sum] = arr[i];
    	        continue;
    	    }
    	    int temp = map[sum]+arr[i];
    	    ans = max(ans,temp);
    	    map[sum] = max(map[sum],arr[i]);
    	}
    	return ans;
    }   
};
